### PR TITLE
Added Chorusfruit teleportaion manipulation

### PR DESCRIPTION
**Pull request by mariusdkm**
This adds the ability to manipulate the destination of a chorusfruit teleportation.

(Inspired by [Are Droppers Actually Random?](https://www.youtube.com/watch?v=Vip25Tet4tA))

**Features**:
- New temprule `chorusManipulation`
- New command `cchorus`
- Seed changes from eating can be calculated, even not fully eaten stuff

With the command `cchorus` a teleportation goal can be set.
That goal is either `relative` or `absolute` to the players position.
The goal can either be an `area` or a `block`.
If a `block` as goal is selected there is another option `--perfectly` this ensures,
that the players hitbox doesn't extend over the blocks boundingbox, so that a player can teleport inside a 1x1 area. 

A block goal is very "improbable" so it requiers a lot of items to be thrown(around 1000+) 
but the chance get's increased the higher the player is standing and the more air is above the block.

**Issues**:
- [x] Since any teleportation above the goal area gets mapped down, the upper y bound is currently ignored, to increase the possibility to land on the desired location. One would have to check every block above the location if that is air.
- [x] If a teleportation fails(it isn't possible to teleport the player to the "random" location),
 the games tries to find a new valid location, which changes the random seed. I can't detect if a teleportation fails, so also the seed change.

- [x] All the text is not translatable
- [ ] All items, except chorus fruits candidates to be thrown, maybe it should be possible to add a blacklist or whitelist?

- [ ] **Not going to change**: When selecting a goal and not using a double, the game defaults to the middle of the block( `8 2 3 -> 8.5 2 3.5` ) 

I would like to highlight the goal somehow, but I have no idea how to do that,
maybe it is possible to use the spectral arrow effect, or the structure block area box?
I coundn't find any function that does that easily.

If there are any structural flaws/issues or generally issues with my code, please let me know,
it's my first large pull request...(you can also contact me on Discord, my tag is `NotSomeBot#1648`)

### CODE REVIEW

File: src/main/java/net/earthcomputer/clientcommands/TempRules.java [hunk 1]
```diff
@@ -53,6 +54,15 @@ public static void setMaxEnchantItemThrows(int maxEnchantItemThrows) {
...
     public static boolean infiniteTools = false;
 
+    @Rule
+    public static boolean chorusManipulation = false;
+
+    @Rule(readOnly = true)
+    //If the goal is relative to the player
+    public static boolean chorusRelativeTel;
+    public static Vec3d chorusGoalV1;
+    public static Vec3d chorusGoalV2;
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/TempRules.java]: Don't put additional variables in `TempRules`. If you can't think of anywhere to put it, you can make a new class inside the `features` package.

File: src/main/java/net/earthcomputer/clientcommands/command/ChorusCommand.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                        .then(blockThen(true))))
+                .then(literal("absolute")
+                    .then(literal("area")
+                        .then(areaThen(false)))
+                    .then(literal("block")
+                        .then(blockThen(false))))
+        ));
+    }
+
+    public static <T> com.mojang.brigadier.builder.RequiredArgumentBuilder<ServerCommandSource, net.minecraft.command.argument.PosArgument> areaThen(boolean relative) {
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/ChorusCommand.java]: Use import statements rather than fully qualified class names.

File: src/main/java/net/earthcomputer/clientcommands/command/ChorusCommand.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        return argument("pos1", vec3())
+                .then(argument("pos2", vec3())
+                    .executes(ctx -> setGoal(getVec3(ctx, "pos1"), getVec3(ctx, "pos2"), relative)));
+    }
+
+    public static <T> com.mojang.brigadier.builder.RequiredArgumentBuilder<ServerCommandSource, net.minecraft.command.argument.PosArgument> blockThen(boolean relative) {
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/ChorusCommand.java]: Import statements

File: src/main/java/net/earthcomputer/clientcommands/command/ChorusCommand.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                sendError(new LiteralText("Goal is to far away!"));
+                return false;
+            }
+        }
+
+        final Vec3d GoalV1 = new Vec3d(Math.min(tempGoalV1.getX(), tempGoalV2.getX()), Math.min(tempGoalV1.getY(), tempGoalV2.getY()), Math.min(tempGoalV1.getZ(), tempGoalV2.getZ()));
+        final Vec3d GoalV2 = new Vec3d(Math.max(tempGoalV1.getX(), tempGoalV2.getX()), Math.max(tempGoalV1.getY(), tempGoalV2.getY()), Math.max(tempGoalV1.getZ(), tempGoalV2.getZ()));
+
+        return throwItemsUntil(rand -> {
+            if (particleCount != 16)
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/ChorusCommand.java]: All if statements should have braces after them. There is some old code in clientcommands that doesn't follow this rule, but all new code should adhere to this.

File: src/main/java/net/earthcomputer/clientcommands/command/ChorusCommand.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                //159 - (7-(itemUseTimeLeft/4)) * 18 = 33 + 4.5 * itemUseTimeLeft
+                for (int i = 0; i < 33 + 4.5 * itemUseTimeLeft; i++)
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/ChorusCommand.java]: Braces

File: src/main/java/net/earthcomputer/clientcommands/command/ChorusCommand.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+
+            double x = (rand.nextDouble() - 0.5D) * 16.0D + pos.getX();
+            double y = rand.nextInt(16) - 8 + pos.getY();
+            double z = (rand.nextDouble() - 0.5D) * 16.0D + pos.getZ();
+
+            if (GoalV1.getX() < x && x < GoalV2.getX() &&
+                    GoalV1.getY() < y && //The top coordinate is special... TODO: Check for things above
+                    GoalV1.getZ() < z && z < GoalV2.getZ()) {
+                if (itemUseTimeLeft == 24) // || itemUseTimeLeft == 24
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/ChorusCommand.java]: Braces

File: src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java [hunk 7]
```diff
@@ -110,8 +114,19 @@ public static void onDrink() {
         //resetCracker("drink");
     }
 
-    public static void onEat() {
-        resetCracker("food");
+    public static void onEat(ItemStack stack, Vec3d pos, int particleCount, int itemUseTimeLeft) {
+        if(canMaintainPlayerRNG()) {
+            //Every time a person eats, the particles are random, and when finished more particles spawn(16)
+            for(int i = 0; i < particleCount * 3 + 3; i++)
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]: Braces

File: src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java [hunk 8]
```diff
@@ -110,8 +114,19 @@ public static void onDrink() {
...
+                nextInt();
+            if (TempRules.chorusManipulation && stack.getItem().getTranslationKey().equals("item.minecraft.chorus_fruit")) {
+                ChorusCommand.onEat(pos, particleCount, itemUseTimeLeft);
+                if(particleCount == 16) //Consumption randoms
+                    for (int i = 0; i < 5; i++)
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]: Braces

File: src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java [hunk 9]
```diff
@@ -110,8 +114,19 @@ public static void onDrink() {
...
+                        nextInt();
+            }
+        } else
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]: Braces

**Review by Earthcomputer — CHANGES_REQUESTED** [review]
Nice feature! Other than the code style:

**Comment by mariusdkm** [comment]
The game doesn't check, if a food regenerates the player, this somehow changes the player seed, so here one may expierience inconsistentcies.

File: src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java [hunk 10]
```diff
@@ -295,7 +316,7 @@ public static boolean throwItem() {
     public static Slot getBestItemThrowSlot(List<Slot> slots) {
         Map<Item, Integer> itemCounts = new HashMap<>();
         for (Slot slot : slots) {
-            if (slot.hasStack() && EnchantmentHelper.getLevel(Enchantments.BINDING_CURSE, slot.getStack()) == 0) {
+            if (slot.hasStack() && EnchantmentHelper.getLevel(Enchantments.BINDING_CURSE, slot.getStack()) == 0 && !slot.getStack().getItem().getTranslationKey().equals("item.minecraft.chorus_fruit")) {
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]:
```suggestion
            if (slot.hasStack() && EnchantmentHelper.getLevel(Enchantments.BINDING_CURSE, slot.getStack()) == 0 && slot.getStack().getItem() != Items.CHORUS_FRUIT) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java [hunk 11]
```diff
@@ -110,8 +113,25 @@ public static void onDrink() {
...
-    public static void onEat() {
-        resetCracker("food");
+    public static void onEat(ItemStack stack, Vec3d pos, int particleCount, int itemUseTimeLeft) {
+        if(canMaintainPlayerRNG()) {
+            //Every time a person eats, the particles are random, and when finished more particles spawn(16)
+            for(int i = 0; i < particleCount * 3 + 3; i++) {
+                nextInt();
+            }
+
+            if (TempRules.chorusManipulation && stack.getItem().getTranslationKey().equals("item.minecraft.chorus_fruit")) {
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]:
```suggestion
            if (TempRules.chorusManipulation && stack.getItem() == Items.CHORUS_FRUIT) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java [hunk 12]
```diff
@@ -267,9 +287,10 @@ public static boolean throwItemsUntil(Predicate<Random> condition, int max) {
                 seed = (seed * MULTIPLIER + ADDEND) & MASK;
             rand.setSeed(seed ^ MULTIPLIER);
         }
-        if (itemsNeeded > max)
+        if (itemsNeeded > max) {
+            sendError(new LiteralText("Condition not possible or would require more than " + max + " item throws"));
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]: Don't put an error message in here, it can be called for reasons other than user interaction. If you want an error message, check the return value.

File: src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java [hunk 13]
```diff
@@ -250,29 +271,40 @@ private static boolean canMaintainPlayerRNG() {
...
+            }
             return false;
-
+        }
         for (int i = 0; i < itemsNeeded; i++) {
-            if (!throwItem())
+            if (!throwItem()) {
+                if (sendErrors) {
+                    sendError(new TranslatableText("playerManip.notEnoughItems", i, itemsNeeded));
+                }
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]: Why do you need these errors now if you are already emitting an error in `ChorusManipulation.onEat`?

File: src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java [hunk 14]
```diff
@@ -22,12 +23,15 @@
...
 import net.minecraft.util.Formatting;
 import net.minecraft.util.StringIdentifiable;
+import net.minecraft.util.math.Vec3d;
 
 import java.lang.reflect.Field;
 import java.util.*;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Predicate;
 
+import static net.earthcomputer.clientcommands.command.ClientCommandManager.sendError;
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]: Don't use static imports except in commands

File: src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java [hunk 15]
```diff
@@ -49,13 +53,13 @@ public static int nextInt() {
...
         if ((bound & -bound) == bound)
-            return (int) ((bound * (long)next(31)) >> 31);
+            return (int) ((bound * (long) next(31)) >> 31);
 
         int bits, val;
         do {
             bits = next(31);
             val = bits % bound;
-        } while (bits - val + (bound-1) < 0);
+        } while (bits - val + (bound - 1) < 0);
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]: Please don't change the formatting in unrelated code

File: src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java [hunk 16]
```diff
@@ -198,8 +219,8 @@ public static void onItemDamage(int amount, LivingEntity holder, ItemStack stack
 
                     if (TempRules.toolBreakWarning && stack.getDamage() + amount >= stack.getMaxDamage() - 30) {
 
-                        if(stack.getDamage() + amount >= stack.getMaxDamage() - 15){
-                            MinecraftClient.getInstance().player.playSound(SoundEvents.ENTITY_EXPERIENCE_ORB_PICKUP, 10,0.1f);
+                        if (stack.getDamage() + amount >= stack.getMaxDamage() - 15) {
+                            MinecraftClient.getInstance().player.playSound(SoundEvents.ENTITY_EXPERIENCE_ORB_PICKUP, 10, 0.1f);
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]: Same here

File: src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java [hunk 17]
```diff
@@ -250,29 +271,40 @@ private static boolean canMaintainPlayerRNG() {
         }
     }
 
-
     // ===== UTILITIES ===== //
-
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]: Same here

**Review by Earthcomputer — CHANGES_REQUESTED** [review]
Also, just to confirm, are you going to address your final two checkboxes?

File: src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java (hunk 13 above)
Comment (mariusdkm) [reply src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]: For detailed feedback.
There can be two reasons a teleport fails. And I think for the end user it is helpfull, to know why he isn't able to manipulate his location.

**Comment by mariusdkm** [comment]
I won't adress the issue with the location input.
But if you think the other would be a nice thing to have, I could implement that.
So my question to you is, do you think a black or whitelist would be good, and should it be inside tempRule or inside the chorusManipulation?

File: src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java (hunk 13 above)
Comment (Earthcomputer) [reply src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]: Make an enum with an "OK" result and other values for possible types of failures. Something similar is done for enchanting manipulation.

Comment (mariusdkm) [reply src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]: So, throwItemUntil should return an Enum?

Comment (Earthcomputer) [reply src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]: Yes

File: src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java [hunk 18]
```diff
@@ -337,6 +357,39 @@ public static long getSeed(Random rand) {
...
+        ThrowItemsState(Boolean success, TranslatableText message) {
+            this.success = success;
+            this.message = message;
+        }
+
+        public TranslatableText getMessage() {
+            return this.message;
+        }
+
+        public ThrowItemsState setMessageArgs(Object... args) {
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java]: Enums shouldn't be mutable. If you need to store any state, use a non-enum class. Sorry, I did not realize you wanted to store state. (Use 4 spaces to indent ofc, I just didn't want to break my spacebar key typing code in github)
```java
public static class ThrowItemsResult {
  private final Type type;
  private final TranslatableText message;

  public ThrowItemsResult(Type type, Object... args) {
    this.type = type;
    this.message = new TranslatableText(type.getTranslationKey(), args);
  }

  public Type getType() {
    return type;
  }

  public TranslatableText getMessage() {
    return message;
  }

  public enum Type {
    NOT_ENOUGH_ITEMS(false, "playerManip.notEnoughItems"),
    NOT_POSSIBLE(false, "playerManip.throwError"),
    UNKNOWN_SEED(false, "commands.cenchant.uncracked"),
    SUCCESS(true, null),
    ;

    private final boolean success;
    private final String translationKey;

    Type(boolean success, String translationKey) {
      this.success = success;
      this.translationKey = translationKey;
    }

    public boolean isSuccess() {
      return success;
    }

    public String getTranslationKey() {
      return translationKey;
    }
  }
}
```

File: src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java [hunk 19]
```diff
@@ -250,32 +269,33 @@ private static boolean canMaintainPlayerRNG() {
         }
     }
 
-
     // ===== UTILITIES ===== //
-
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java:255]: This formatting change still needs reverting. This comment is a section heading, and the whitespace is needed for clarity, and to show that the comment does not merely apply to the next method.

**Review by Earthcomputer — CHANGES_REQUESTED** [review]
Very nearly ready to approve

**Review by Earthcomputer — APPROVED** [review]
Thank you!
